cpp/test/cpu_operations_test/trace_test.cc

```diff
@@ -28,13 +28,25 @@
 #include "include/cpu_operations.h"
 #include "include/matrix.h"
 
-TEST(Trace, TraceMatrix) {
-Eigen::MatrixXi a(4, 4);
-a << 8, 5, 3, 4,
-     2, 4, 8, 9,
-     7, 6, 1, 0,
-     9, 2, 5, 7;
-int correct_ans = 20;
-int calc_ans = Nice::CpuOperations<int>::Trace(a);
-EXPECT_EQ(correct_ans, calc_ans);
+template<class T>
+class TraceTest : public ::testing::Test {
+  public:
+  Nice::Matrix<T> m1;
+
+  int Tracer() {
+    return Nice::CpuOperations<T>::Trace(m1);
+  }
+};
+
+typedef ::testing::Types<int, float, double> MyTypes;
+TYPED_TEST_CASE(TraceTest, MyTypes);
+
+TYPED_TEST(TraceTest, BasicTest) {
```

Comment:
Since the correct_ans is always an int but the value of Nice::CpuOperations<T>::Trace(m1) is of type T, I am a little surprised that this would pass the EXPECT_EQ() test, could you check if this->Tracer() returns a data type T?